Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+          symbols.StringJoinMethodSymbols,
+          s => SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, s),
+          out var methodSymbol))
+      {
+        var argumentTypes = GetArgumentTypeSymbols (node, ctx.SemanticModel);
+        if (AnyWithResult (
+            argumentTypes,
+            a => SymbolEqualityComparer.Default.Equals (a, symbols.WebStringTypeSymbol)
+                 || SymbolEqualityComparer.Default.Equals (a, symbols.PlainTextStringTypeSymbol)
+                 || ctx.Compilation.ClassifyConversion (a, symbols.WebStringEnumerable).IsImplicit
+                 || ctx.Compilation.ClassifyConversion (a, symbols.PlainTextStringEnumerable).IsImplicit,
+            out var result))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              methodSymbol.ToString(),
+              result.ToString(),
+              $"Encode the {result.ToString()} instances first.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+    }
+
+    private static bool AnyWithResult<TSymbol> (IEnumerable<TSymbol> methods, Func<TSymbol, bool> predicate, out TSymbol match)
+        where TSymbol : ISymbol
+    {
+      match = methods.FirstOrDefault (predicate)!;
+      return match != null;
+    }
+
+    private static ITypeSymbol? GetFirstArgumentTypeSymbol (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      var firstArgumentExpression = node.ArgumentList.Arguments.First().Expression;
+      return model.GetTypeInfo (firstArgumentExpression).Type;
+    }
+
+    private static IReadOnlyCollection<ITypeSymbol> GetArgumentTypeSymbols (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      return node.ArgumentList.Arguments.Select (s => model.GetTypeInfo (s.Expression).Type).OfType<ITypeSymbol>().ToList();
+    }
+
+    private static IReadOnlyCollection<IMethodSymbol> GetStringJoinMethodSymbols (INamedTypeSymbol stringSymbol)
+    {
+      return stringSymbol
+          .GetMembers ("Join")
+          .Cast<IMethodSymbol>()
+          .ToList();
+    }
+
+    private static IMethodSymbol GetHtmlTextWriterWriteObjectMethodSymbol (INamedTypeSymbol htmlTextWriterSymbol, INamedTypeSymbol? objectSymbol)
+    {
+      return htmlTextWriterSymbol
+          .GetMembers ("Write")
+          .Cast<IMethodSymbol>()
+          .Single (m => m.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals (m.Parameters.Single().Type, objectSymbol));
+    }
+
+    private static IMethodSymbol GetStringBuilderAppendObjectMethodSymbol (INamedTypeSymbol stringBuilderTypeSymbol, INamedTypeSymbol? objectSymbol)
+    {
+      return stringBuilderTypeSymbol
+          .GetMembers ("Append")
+          .Cast<IMethodSymbol>()
+          .Single (m => m.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals (m.Parameters.Single().Type, objectSymbol));
+    }
+
+    class SymbolContext
```

Comment:
Also, our style guide says that nested types are placed on the top of the class (nested types, const, static, instance fields, ctor, members)